source/postprocess/geoid.cc

```diff
@@ -259,71 +249,155 @@ namespace aspect
-            fe_face_values[this->introspection().extractors.temperature].get_function_values(topo_vector, topo_values);
 
-            // if the cell at top boundary, add its contributions dynamic topography storage vector
-            if (face_idx != numbers::invalid_unsigned_int && at_upper_surface)
+            if (include_surface_topo_contribution == true)
+              {
+                if (use_real_surface_topography == true)
+                  {
+                    // If the cell is at the top boundary, add its contributions to the topography surface storage vector.
+                    if (face_idx != numbers::invalid_unsigned_int && at_upper_surface)
+                      {
+                        for (unsigned int q=0; q<fe_face_values.n_quadrature_points; ++q)
+                          {
+                            const Point<3> current_position = fe_face_values.quadrature_point(q);
+                            const double topography = this->get_geometry_model().height_above_reference_surface(current_position);
+                            surface_stored_values.emplace_back (std::make_pair(current_position, std::make_pair(fe_face_values.JxW(q), topography)));
+                          }
+                      }
+                  }
+                else
+                  {
+                    // Get a pointer to the dynamic topography postprocessor.
+                    const Postprocess::DynamicTopography<3> &dynamic_topography =
+                      this->get_postprocess_manager().template get_matching_postprocessor<Postprocess::DynamicTopography<3> >();
+
+                    // Get the already-computed dynamic topography solution.
+                    const LinearAlgebra::BlockVector &topo_vector = dynamic_topography.topography_vector();
+
+                    std::vector<double> topo_values( quadrature_formula_face.size());
+
+                    fe_face_values[this->introspection().extractors.temperature].get_function_values(topo_vector, topo_values);
+
+                    // If the cell is at the top boundary, add its contributions to the topography surface storage vector.
+                    if (face_idx != numbers::invalid_unsigned_int && at_upper_surface)
+                      {
+                        for (unsigned int q=0; q<fe_face_values.n_quadrature_points; ++q)
+                          surface_stored_values.emplace_back (std::make_pair(fe_face_values.quadrature_point(q), std::make_pair(fe_face_values.JxW(q), topo_values[q])));
+                      }
+                  }
+              }
+            else
               {
-                for (unsigned int q=0; q<fe_face_values.n_quadrature_points; ++q)
-                  surface_stored_values.emplace_back (std::make_pair(fe_face_values.quadrature_point(q), std::make_pair(fe_face_values.JxW(q), topo_values[q])));
+                // If the cell is at the top boundary, add its contributions to the topography surface storage vector.
```

Comment:
Can you adjust this comment to reflect what you're doing in the next couple of lines?